clang/test/SemaSYCL/attr-add-ir-attributes-merge.cpp

```diff
@@ -0,0 +1,50 @@
+// RUN: %clang_cc1 -fsycl-is-device -std=gnu++11 -fsyntax-only -verify %s
+
+void FunctionRedecl1();
+[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void FunctionRedecl1();
+[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void FunctionRedecl1();
+[[__sycl_detail__::add_ir_attributes_function("Attr2", "Attr1", true, 1)]] void FunctionRedecl1(); // expected-note {{conflicting attribute is here}}
+[[__sycl_detail__::add_ir_attributes_function("Attr3", false)]] void FunctionRedecl1(){};          // expected-error {{attribute 'add_ir_attributes_function' is already applied with different arguments}}
+
+[[__sycl_detail__::add_ir_attributes_function("Attr1", "Attr2", 1, true)]] void FunctionRedecl2();
+void FunctionRedecl2();
+
+void FunctionRedecl3();
+[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr1", "Attr2", 1, true)]] void FunctionRedecl3();
+[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr1", "Attr2", 1, true)]] void FunctionRedecl3();
+[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr2", "Attr1", true, 1)]] void FunctionRedecl3();
+[[__sycl_detail__::add_ir_attributes_function({"Attr3", "Attr1"}, "Attr1", "Attr2", 1, true)]] void FunctionRedecl3(); // expected-note {{conflicting attribute is here}}
+[[__sycl_detail__::add_ir_attributes_function({"Attr1"}, "Attr1", "Attr2", 1, true)]] void FunctionRedecl3();          // expected-error {{attribute 'add_ir_attributes_function' is already applied with different arguments}} expected-note {{conflicting attribute is here}}
```

Comment:
Sorry if we've already discussed this but technically this case should be ok right? Since functionally 
[[__sycl_detail__::add_ir_attributes_function({"Attr1", "Attr3"}, "Attr1", "Attr2", 1, true)]] void FunctionRedecl3() = [[__sycl_detail__::add_ir_attributes_function("Attr1", 1)]] void FunctionRedecl3();

Reply:
Technically, yes. However, to do that we would need to inspect the values in the filter and compare them to the attribute names early on. It should not be too difficult to do, given that we already extract the values at the time of "merging", but I would prefer keeping the merging logic simple and somewhat restrictive for now until we know exactly what behavior we need from it in the future.

Reply:
Sounds good to me. Could you file an issue so we don't lose track of it?

Reply:
I have opened an internal issue to keep track of decisions made regarding the merging logic for these attributes.

Reply:
Thanks. I also noticed another issue which FE should look into at some point. It looks like incorrect attributes are still attached to redeclarations (after generating error) but I don't understand why because your code looks correct. We may have to call `dropAttr()` instead of returning `nullptr ` when we merge invalid attributes. I'm not sure though and I think this issue probably affects many existing attributes, and so should be done as a follow-up (not necessarily by you unless you want to :D) fixing all attributes.